[Step3]질문 삭제하기 리팩터링 PR 요청 드립니다.

re-mora opened the pull request:
안녕하세요 :) 교육생 최용석입니다.

3단계 질문 삭제하기 리팩터링 미션 구현이 완료되어, 코드 리뷰 요청드립니다.

이번 미션 역시 미흡한 부분이 있다면 피드백 부탁드립니다 :)

--- 

이번 PR에도 step2에서 피드백 주셨던 내용에 대한 TC를 아래 커밋에 작성하였습니다.

- [FetchType.LAZY VS FetchType.EAGER 동작 비교](https://github.com/next-step/jwp-qna/commit/89df5204a78d3ba00ac02ddf15a06bb4d04a9def)
- [JPQL의 Join VS Fetch Join 동작 비교](https://github.com/next-step/jwp-qna/commit/f68a2b12b319d8fdfc23727a0341bd142b633e61)
- [Fetch Join VS@EntityGraph 동작 비교(inner join vs left join)](https://github.com/next-step/jwp-qna/commit/b84da39009c093e1768903744d06bbaee72eba1c)
- [N+1이 문제와 N+1에 대한 해결 방안](https://github.com/next-step/jwp-qna/commit/141c200d4b0f59f5fb549e1b33c9537d9b4dd0b4)
- [트랜잭션에서 밖에서 프록시 객체 초기화 시 발생하는 LazyInitializationException 검증](https://github.com/next-step/jwp-qna/commit/0dfe63e43f4f3bef1eb03b3338cd7b4d3273129b)

한번 더 생각할 수 있도록 리뷰 해주신 덕분에, 이번 미션을 진행하는 동안 전반적으로 JPA에 대해 돌아보는 계기가 되었습니다 :) 
길지 않은 기간 이었지만 리뷰 해주시느라 고생 많으셨습니다. 감사합니다!

re-mora committed:
test: step2 코드리뷰 피드백 예제를 위한 Entity 작성

re-mora committed:
docs: step2 코드 리뷰 피드백 내용 정리

re-mora committed:
test: 코드 리뷰 피드백 관련 TC 설명 수정

re-mora committed:
test: 코드 리뷰 피드백 - FetchType.LAZY VS FetchType.EAGER 동작 비교

- FetchType.EAGER 전략으로 인한 연관 Entity의 즉시 로딩 여부 검증
- FetchType.LAZY 전략으로 인한 연관 Entity의 프록시 객체 여부 검증

re-mora committed:
test: 코드 리뷰 피드백 - JPQL의 Join VS Fetch Join 동작 비교

- JPQL-Join : Select 절의 대상 Entity만 조회, 연관된 Entity의 프록시 객체 여부 검증
- JPQL-Fetch Join : Entity 조회 시 연관 Entity의 즉시 로딩 여부 검증

re-mora committed:
test: Fetch Join VS @EntityGraph 동작 비교(inner join vs left join)

- JPQL-Fetch Join : Inner Join 동작 검증
- Spring Data JPA-@EntityGraph : Left Outer Join 동작 검증
- JPQL-Fetch Join : `distinct`를 이용하여 카테시안 곱 연산으로 인해 발생하는 중복 문제 해결

re-mora committed:
test: 코드 리뷰 피드백 - N+1이 문제와 N+1에 대한 해결 방안

- FetchType.EAGER의 N+1 발생 여부 검증
- FetchType.Lazy의 N+1 발생 여부 검증
- @EntityGraph의 attributePaths 속성을 이용한 N+1 문제 해결
- Fetch Join을 이용한 N+1 문제 해결

re-mora committed:
test: 코드 리뷰 피드백 - 트랜잭션에서 밖에서 프록시 객체 초기화 시 발생하는 LazyInitializationException 검증

re-mora committed:
docs: step3 요구사항 분석 및 구현 기능 목록 작성

re-mora committed:
refactor: TC 작성에 반복적으로 사용되는 질문 및 답변 객체 생성 부 수정

TC를 위해 매번 파라미터를 입력하는 부분을 Counter로 처리하여 여러 객체 생성에 용이 하도록 수정

re-mora committed:
feat: 질문 삭제 기능 구현

질문 삭제 요청자와 질문 작성자, 답변 작성자에 대한 유효성 검증 추가

질문 삭제 관련 TC 작성
- 질문 삭제 요청자와 질문 작성자가 다른 경우 예외 발생 검증 TC 작성
- 질문 작성자와 답변 작성자가 다른 경우 예외 발생 검증 TC 작성

기능 구현에 따른 TODO List 수정 및 갱신

re-mora committed:
feat: 삭제 이력 생성 기능 구현

- 삭제 이력 생성
- 답변이 없는 질문 삭제 시, 삭제 이력 생성

기능 구현에 따른 TODO List 수정 및 갱신

re-mora committed:
refactor: 삭제 이력 객체를 일급 컬렉션으로 구현

- 질문 객체의 답변 존재 여부에 따라 삭제 이력 생성 처리 구현
- 일급 컬렉션 구현에 따른 TC 수정

기능 구현에 따른 TODO List 수정 및 갱신

re-mora committed:
refactor: 답변 목록을 일급 컬렉션으로 구현

- 질문 삭제 시, 해당 질문의 답변 목록 일괄 삭제 처리 구현
- 질문 작성자와 답변 작성자가 같은지 여부를 판별하는 메서드 구현

기능 구현에 따른 TODO List 수정 및 갱신

re-mora committed:
chore(re-package): 패키지 구조 정리 (domain과 repository를 구분)

re-mora committed:
refactor: 서비스 계층의 비지니스 로직을 연관 도메인 계층으로 분산

권한 관련 예외 발생 시, 잘못된 예외 처리 수정
  - AS-IS : UnAuthorizedException
  - TO-BE : CannotDeleteException

re-mora committed:
refactor: `Question`의 답변 목록 접근 시, 삭제되지 않은 답변 조회로 변경

필드 레벨에서 @Where를 이용하여 질문 조회 시, 삭제되지 않은 질문 목록이 LazyLoading 되도록 설정

re-mora committed:
chore(polish): 사용하지 않는 의존성 및 getter 메서드 삭제

불필요 TC 삭제

seokhongkim commented on docs/step2-README.md:
```diff
@@ -105,3 +105,42 @@ DB 컬럼 중심적인 매핑에서 객체 간 연관관계 매핑으로 변경
+  - 다른쪽의 toString에서 참조를 끊어냄으로써 순환참조를 방지
+
+> N+1 문제와 이를 해결하기 위한 방법
+- Fetch Join을 이용하여 연관관계를 가지는 Entity를 함께 조회
```
피드백 드린 내용을 꼼꼼하게 잘 학습해주셔서 피드백 드린 보람이 더 크게 느껴지네요! 감사합니다. :)

fetch join 을 사용할 때 limit 쿼리와 함께 사용하면 어떤 일이 발생하는지도 같이 보셔도 도움이 될 것 같습니다.

예를 들면, 질문 5개와 거기에 달린 답변을 모두 조회하고 싶을 때 fetch join + distinct + limit 5 를 사용하면 어떻게 동작할까요? distinct 없이 사용하는 경우에는 어떻게 동작할까요?

re-mora replied:
말씀하셨던 내용에 대한 학습테스트를 아래 커밋에 작성하였습니다!

[test: 1:N 연관관계의 fetch join 사용 시, 주의 사항 및 해결 방안에 대한 학습 테스트](https://github.com/next-step/jwp-qna/pull/565/commits/7d0fdf566716757c630b693f8844d25a7b2bee07)

다만  distinct 유무에 따른 아래 두가지 학습 테스트를 작성 하였지만, 큰 차이점을 발견하지 못했습니다. 
- fetch join + distinct + limit 5 
- fetch join + limit 5 

distinct 유무에 따라 중복 제거에 대한 부분에 차이가 있을 것이라고 예상 하였지만, distinct를 명시하지 않은 페이징 쿼리의 결과에 중복이 제거되어 있어서 이 부분에 대해 제가 추가로 알아야 할 사항이 있다면 한번 더 피드백 부탁드릴 수 있을까요?

seokhongkim replied:
데이터 개수에 따라서 테스트 결과가 달라질 수 있는데, 이 부분을 제가 말씀드리지 않았네요. 😓 
[블로그](https://tecoble.techcourse.co.kr/post/2020-10-21-jpa-fetch-join-paging/)에 올라온 내용을 읽어보시고, 테스트에 참고하시면 도움이 될 것 같습니다. :)

seokhongkim commented on src/main/java/qna/domain/Answer.java:
```diff
@@ -76,6 +76,10 @@ public void toQuestion(Question question) {
         this.question = question;
     }
 
+    public void delete() {
```
`Answer` 라는 도메인의 입장에서 봤을 때, 질문은 그대로 남아있고 답변만 삭제하는 경우 `delete` 는 어떻게 동작해야 할까요?

hint: 답변을 작성한 사람만 삭제할 수 있어야 한다는 제약조건은 어느 도메인에서 담당해야 할 역할일까요?

re-mora replied:
질문과 답변을 모두 삭제하는 경우만 고려하여 작성자에 대한 유효성 검사를 답변 일급 컬렉션에 작성하게 되었습니다.
답변만 삭제하는 경우를 고려하여, 작성자에 대한 유효성 검사부의 위치를 수정하겠습니다!

seokhongkim commented on src/main/java/qna/domain/Answers.java:
```diff
@@ -0,0 +1,31 @@
+
+    public void deleteAll(User questionWriter) throws CannotDeleteException {
+        if (containsNotEqualsWriter(questionWriter)) {
+            throw new CannotDeleteException("다른 사람이 쓴 답변이 있어 삭제할 수 없습니다.");
```
질문을 삭제할 때 발생할 수 있는 예외 내용으로 보이네요!
답변을 관리하는 컬렉션에서 질문 도메인에 대해 자세히 알고 있는 것은 아닐까요?

re-mora replied:
이 부분도 연관도가 높은 각 도메인으로 묶일 수 있도록 위치를 수정하겠습니다!

seokhongkim commented on src/main/java/qna/repository/AnswerRepository.java:
```diff
@@ -1,11 +1,11 @@
 public interface AnswerRepository extends JpaRepository<Answer, Long> {
-    List<Answer> findByQuestionAndDeletedFalse(Question question);
-
     Optional<Answer> findByIdAndDeletedFalse(Long id);
```
이 메소드를 사용하는 곳이 남아 있을까요?

seokhongkim commented on src/main/java/qna/service/QnaService.java:
```diff
@@ -6,53 +6,34 @@
-        }
-        deleteHistoryService.saveAll(deleteHistories);
+        question.delete(loginUser);
+        DeleteHistories deleteHistories = DeleteHistories.of(question);
```
삭제를 한 후 `DeleteHistory` 를 만들어줘야 한다는 것을 서비스에서 관리하고 있네요!
`Question`, `Answer` 도메인에서 데이터가 삭제되는 경우 삭제 이력을 반환하도록 만들어 보는 것은 어떨까요?

re-mora replied:
Question/Answer 도메인의 반환형을 다른 도메인으로 구성하는 것이 어색하여 따로 분리하게 되었습니다.
말씀하신 것처럼 삭제 로직 수행 결과로 삭제 이력을 반환할 수 있도록 구성 하겠습니다!

seokhongkim commented on src/test/java/qna/feedback/Ex02_Lazy_Type_And_Eager_Type_Test.java:
```diff
@@ -0,0 +1,101 @@
+@TestConstructor(autowireMode = AutowireMode.ALL)
+@Import({TeamGenerator.class, MemberGenerator.class, ClassGenerator.class, StudentGenerator.class})
+@DisplayName("코드 리뷰 피드백 : FetchType.LAZY VS FetchType.EAGER 동작 비교하기")
+public class Ex02_Lazy_Type_And_Eager_Type_Test {
```
학습 테스트 좋습니다! 👍

seokhongkim reviewed (changes requested):
용석님, JPA 3단계 미션도 잘 진행해주셨습니다.

2단계 미션에서 드린 피드백에 대해 학습 테스트를 모두 작성하며, 자세히 학습하신 것 같네요!
이번 단계 미션도 꼼꼼한 테스트와 함께 작은 커밋을 유지하며 잘 진행해주셨습니다. 👍 

코멘트를 조금 남겼으니 확인 부탁드립니다. 🙏

re-mora committed:
docs: step3 코드 리뷰 피드백 TODO List 정리

re-mora committed:
chore(polish): TC 에서만 참조하는 쿼리 메서드 삭제

re-mora committed:
refactor: 답변만 삭제 하는 경우, 작성자에 대한 누락된 유효성 검사 추가

작성자에 대한 유효성검사를 위해 파라미터 추가
  - AS-IS : delete()
  - TO-BE : delete(User loginUser)

일급컬렉션에 위치했던 연관성이 떨어지는 질문관련 유효성 검사 제거

re-mora committed:
refactor: 서비스 계층의 삭제 이력 생성 비지니스 로직을 각 도메인 계층으로 분산

- 삭제 로직에 삭제 이력 반환부를 추가함에 따라 관련된 TC에 삭제 이력에 대한 검증 부 추가
- 서비스 계층의 삭제 이령 생성 비지니스 로직을 각 도메인 계층에 분산함에 따라
역할이 모호해진 삭제 이력 일급 컬렉션 삭제

re-mora committed:
test: 1:N 연관관계의 fetch join 사용 시, 주의 사항 및 해결 방안에 대한 학습 테스트

- 1:N fetch join의 카테이션 곱 연산으로 인한 중복 발생 TC 작성
- distinct를 이용한 1:N fetch join의 중복 해결 TC 작성
- 1:N fetch join의 페이징 처리 시, 메모리 이슈 관련 TC 작성

seokhongkim commented on src/main/java/qna/repository/AnswerRepository.java:
```diff
@@ -1,11 +1,10 @@
-    List<Answer> findByQuestionAndDeletedFalse(Question question);
 
-    Optional<Answer> findByIdAndDeletedFalse(Long id);
+//    Optional<Answer> findByIdAndDeletedFalse(Long id);
```
사용하지 않는 코드는 주석 처리보다는 삭제하시는 것을 권장합니다.
나중에 필요하시면 commit log 를 보시면 어떤 코드인지 찾을 수 있어요!

seokhongkim commented on src/main/java/qna/repository/QuestionRepository.java:
```diff
@@ -0,0 +1,50 @@
+}
+
+
+//글로벌서비스 개발팀이 현재 운영하고 있는 서비스가 있는지
```
코드와 상관 없는 내용이 붙여넣기 된 것 같습니다. :)

re-mora replied:
헉.. 삭제하도록 하겠습니다.. 😭

seokhongkim commented on src/main/java/qna/domain/Question.java:
```diff
@@ -71,6 +74,22 @@ public void addAnswer(Answer answer) {
+            throw new CannotDeleteException("질문을 삭제할 권한이 없습니다.");
+        }
+
+        List<DeleteHistory> deleteHistories = new ArrayList<>();
```
본인이 작성하지 않은 답변이 있는 경우 발생했던 `CannotDeleteException("다른 사람이 쓴 답변이 있어 삭제할 수 없습니다.")` 예외 처리 코드가 사라진 것 같습니다.

re-mora replied:
기존 답변 컬렉션 객체에 구현 했던 코드를 삭제 하면서 누락된 것 같습니다 😭 
질문 삭제 시, 본인이 작성하지 않은 답변이 포함되는 경우에 대한 예외 처리를 아래 커밋에 추가하였습니다.

[refactor: 본인이 작성하지 않은 답변이 포함되는 경우 예외 누락 추가](https://github.com/next-step/jwp-qna/pull/565/commits/a7b4666d503aa32237f675a9e2af9b545e15be2e)

seokhongkim reviewed (changes requested):
용석님, 피드백 잘 반영해주셨습니다! 👍 

질문주신 내용에 대한 답변도 남겼습니다.
간단한 코멘트 조금 남겼으니 확인 부탁드립니다. 🙏

re-mora committed:
chore(polish): 불필요 코드 삭제

re-mora committed:
chore(polish): 미사용 코드 삭제

re-mora committed:
refactor: 본인이 작성하지 않은 답변이 포함되는 경우 예외 누락 추가

seokhongkim reviewed (approved):
용석님, 3단계 미션 잘 마무리해주셨습니다. 👍 

미션을 진행하시며 피드백으로 말씀드린 내용에 대해서도 열심히 학습하신 것 같습니다.
학습하신 내용이 앞으로 JPA 사용하실 때 도움이 되셨으면 좋겠네요!

남은 과정도 많이 학습하시고 성장하실 수 있기를 응원하겠습니다. 💪 
JPA 미션 진행하느라 수고하셨습니다! 👏
